### PR TITLE
Remove all datasets related `extras_require` and move requirements to `pyproject.toml`

### LINKED ISSUE #2127: Remove all datasets related `extras_require=` from `kedro`
### Description
The `extras_require` related to datasets aren't needed anymore for `0.19.0` because Kedro will just use `kedro-datasets`.

### PR DESCRIPTION
**Pull request by ankatiyar**
> **NOTE:** Kedro datasets are moving from `kedro.extras.datasets` to a separate `kedro-datasets` package in
> [`kedro-plugins` repository](https://github.com/kedro-org/kedro-plugins). Any changes to the dataset implementations
> should be done by opening a pull request in that repository.
## Description

Resolve #2127 

## Development notes

- Moved all test and docs requirement to `pyproject.toml` from `setup.py`
- There's some `package_data` stuff left in `setup.py` which can be moved as a follow up and we can remove `setup.py` entirely
- Had to update some tests to make them run without `kedro.extras.datasets`
- `pandas.util.testing` was deprecated, so updated it.
## Checklist

- [ ] Read the [contributing](https://github.com/kedro-org/kedro/blob/main/CONTRIBUTING.md) guidelines
- [ ] Signed off each commit with a [Developer Certificate of Origin (DCO)](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/managing-repository-settings/managing-the-commit-signoff-policy-for-your-repository)
- [ ] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Updated the documentation to reflect the code changes
- [ ] Added a description of this change in the [`RELEASE.md`](https://github.com/kedro-org/kedro/blob/main/RELEASE.md) file
- [ ] Added tests to cover my changes
- [ ] Checked if this change will affect Kedro-Viz, and if so, communicated that with the Viz team

### CODE REVIEW

File: tests/io/test_data_catalog.py [hunk 1]
```diff
@@ -762,16 +762,18 @@ def test_replacing_nonword_characters(self):
...
         monkeypatch.setenv("AWS_SECRET_ACCESS_KEY", "dummmy")
         version = Version(load=None, save=None)
+        mocker.patch("kedro_datasets.pandas.csv_dataset.fsspec")
         versioned_dataset = CSVDataSet("s3://bucket/file.csv", version=version)
         pattern = re.escape(
             f"Did not find any versions for {versioned_dataset}. "
             f"This could be due to insufficient permission."
         )
+
         with pytest.raises(DatasetError, match=pattern):
```
[the remark this answers is not in the document]
Comment (ankatiyar) [reply tests/io/test_data_catalog.py:777]: Without this it was erroring with `The AWS Access Key Id you provided does not exist in our records.` where the `load()` function was actually looking for stuff in the fake bucket.

[the remark this answers is not in the document]
Comment (ankatiyar) [reply tests/io/test_data_catalog.py:777]: We don't need to patch the credentials actually if we're patching `fsspec`

File: pyproject.toml [hunk 2]
```diff
@@ -54,7 +54,56 @@ classifiers = [
...
+    "behave==1.2.6",
+    "blacken-docs==1.9.2",
+    "black~=22.0",
+    "coverage[toml]",
+    "import-linter[toml]==1.8.0",
+    "ipython>=7.31.1, <8.0; python_version < '3.8'",
+    "ipython~=8.10; python_version >= '3.8'",
+    "Jinja2<3.1.0",
+    "jupyterlab_server>=2.11.1, <2.16.0",  # 2.16.0 requires importlib_metedata >= 4.8.3 which conflicts with flake8 requirement
+    "jupyterlab~=3.0, <3.6.0",  # 3.6.0 requires jupyterlab_server~=2.19
```
Comment (merelcht) [inline pyproject.toml:71]: I removed these bounds in https://github.com/kedro-org/kedro/pull/3072/commits/56370d28b3918cd020823d5509cdea20bfbc8afe

**Review by merelcht — COMMENTED** [review]
Generally looks good! Can't wait to see `setup.py` completely removed.

File: pyproject.toml [hunk 3]
```diff
@@ -52,7 +52,57 @@ classifiers = [
...
+dynamic = ["readme", "version"]
+
+[project.optional-dependencies]
+test = [
+    "bandit>=1.6.2, <2.0",
+    "behave==1.2.6",
+    "blacken-docs==1.9.2",
+    "black~=22.0",
+    "coverage[toml]",
+    "fsspec<2023.9", # Temporary, newer version causing "test_no_versions_with_cloud_protocol" to fail
```
Comment (merelcht) [inline pyproject.toml:64]: Would be good to have a follow up ticket to unpin this again.

**Review by merelcht — APPROVED** [review]
LGTM! 👍

File: pyproject.toml (hunk 3 above)
Comment (ankatiyar) [reply pyproject.toml:64]: Yep, I'll create one.

File: setup.py [hunk 4]
```diff
@@ -1,14 +1,7 @@
...
 
-# at least 1.3 to be able to use XMLDataSet and pandas integration with fsspec
-PANDAS = "pandas~=1.3"
-SPARK = "pyspark>=2.2, <3.4"
-HDFS = "hdfs>=2.5.8, <3.0"
-S3FS = "s3fs>=0.3.0, <0.5"
-
 template_files = []
 for pattern in ["**/*", "**/.*", "**/.*/**", "**/.*/.**"]:
     template_files.extend(
```
Comment (astrojuanlu) [inline setup.py:7]: I'm so looking forward to deleting this too 😄 (in a future PR)

**Review by astrojuanlu — APPROVED** [review]
Looks good to me! Thanks @ankatiyar ⭐

**Review by AhdraMeraliQB — APPROVED** [review]

**Pull request merged**